app/scripts/services/oauth/oauth-service.ts

```
@@ -0,0 +1,117 @@
+import { AuthConnection } from '@metamask/seedless-onboarding-controller';
+import { BaseLoginHandler } from './base-login-handler';
+import { createLoginHandler } from './create-login-handler';
+import type {
+  OAuthLoginEnv,
+  OAuthLoginResult,
+  OAuthServiceOptions,
+  WebAuthenticator,
+} from './types';
+
+export default class OAuthService {
+  #env: OAuthLoginEnv;
+
+  #webAuthenticator: WebAuthenticator;
+
+  constructor({ env, webAuthenticator }: OAuthServiceOptions) {
+    this.#env = env;
+    this.#webAuthenticator = webAuthenticator;
+  }
+
+  /**
+   * Start the OAuth login process for the given social login type.
+   *
+   * @param authConnection - The social login type to login with.
+   * @returns The login result.
+   */
+  async startOAuthLogin(
+    authConnection: AuthConnection,
+  ): Promise<OAuthLoginResult> {
+    // get the redirect URI for the OAuth login
+    const redirectUri = this.#webAuthenticator.getRedirectURL();
+
+    // create the login handler for the given social login type
+    // this is to get the Jwt Token in the exchange for the Authorization Code
+    const loginHandler = createLoginHandler(
+      authConnection,
+      redirectUri,
+      this.#env,
+    );
+
+    // launch the web auth flow to get the Authorization Code from the social login provider
+    const redirectUrlFromOAuth = await this.#webAuthenticator.launchWebAuthFlow(
+      {
+        interactive: true,
+        url: loginHandler.getAuthUrl(),
+      },
+    );
```

Above I added a couple comments about the importance of the `state` value. In launchWebAuthFlow's optional [callback](https://developer.chrome.com/docs/extensions/reference/api/identity#method-launchWebAuthFlow) we should validate that the `state` we get back from the oauth provider matches what we provided at the start of the flow. If the value does not match, we should throw an error.